src/raw/serde.rs

```diff
@@ -0,0 +1,527 @@
+            // sees the RAW_DOCUMENT_NEWTYPE but the next type isn't a document.
+            OwnedOrBorrowedRawBson::Borrowed(RawBson::Binary(b))
+                if b.subtype == BinarySubtype::Generic =>
+            {
+                Ok(Self::Borrowed(
+                    RawDocument::from_bytes(b.bytes).map_err(serde::de::Error::custom)?,
+                ))
+            }
+            OwnedOrBorrowedRawBson::Owned(OwnedRawBson::Binary(b))
+                if b.subtype == BinarySubtype::Generic =>
+            {
+                Ok(Self::Owned(
+                    RawDocumentBuf::from_bytes(b.bytes).map_err(serde::de::Error::custom)?,
+                ))
+            }
+
+            o => Err(serde::de::Error::custom(format!(
+                "expected raw document, instead got {:?}",
+                o
+            ))),
+        }
+    }
+}
+
+/// Wrapper type that can deserialize either an owned or a borrowed raw BSON array.
+#[derive(Debug)]
+pub(crate) enum OwnedOrBorrowedRawArray<'a> {
+    Owned(RawArrayBuf),
+    Borrowed(&'a RawArray),
+}
+
+impl<'a> OwnedOrBorrowedRawArray<'a> {
+    pub(crate) fn into_owned(self) -> RawArrayBuf {
+        match self {
+            Self::Owned(o) => o,
+            Self::Borrowed(b) => b.to_owned(),
+        }
+    }
+}
+
+impl<'a, 'de: 'a> Deserialize<'de> for OwnedOrBorrowedRawArray<'a> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        match deserializer
+            .deserialize_newtype_struct(RAW_ARRAY_NEWTYPE, OwnedOrBorrowedRawBsonVisitor)?
+        {
+            OwnedOrBorrowedRawBson::Borrowed(RawBson::Array(d)) => Ok(Self::Borrowed(d)),
+            OwnedOrBorrowedRawBson::Owned(OwnedRawBson::Array(d)) => Ok(Self::Owned(d)),
+
+            // For non-BSON formats, RawArray gets serialized as bytes, so we need to deserialize
+            // from them here too. For BSON, the deserializier will return an error if it
+            // sees the RAW_DOCUMENT_NEWTYPE but the next type isn't a document.
+            OwnedOrBorrowedRawBson::Borrowed(RawBson::Binary(b))
+                if b.subtype == BinarySubtype::Generic =>
+            {
+                let doc = RawDocument::from_bytes(b.bytes).map_err(serde::de::Error::custom)?;
+                Ok(Self::Borrowed(RawArray::from_doc(doc)))
+            }
+            OwnedOrBorrowedRawBson::Owned(OwnedRawBson::Binary(b))
+                if b.subtype == BinarySubtype::Generic =>
+            {
+                let doc = RawDocumentBuf::from_bytes(b.bytes).map_err(serde::de::Error::custom)?;
+                Ok(Self::Owned(RawArrayBuf::from_raw_document_buf(doc)))
+            }
+
+            o => Err(serde::de::Error::custom(format!(
+                "expected raw array, instead got {:?}",
+                o
+            ))),
+        }
+    }
+}
+
+/// A visitor used to deserialize types backed by raw BSON.
+pub(crate) struct OwnedOrBorrowedRawBsonVisitor;
```

Comment:
This is essentially the same as the old `RawBsonVisitor`, but it now can handle owned stuff (e.g. `visit_string`, `visit_byte_buf`) and arbitrary maps and sequences.